packages/authorization-platformstate-writer/src/consumerServiceV2.ts

```diff
@@ -1,19 +1,485 @@
+              {
+                ...clientPurposeEntry,
+                GSIPK_consumerId_eserviceId: makeGSIPKConsumerIdEServiceId({
+                  consumerId: client.consumerId,
+                  eserviceId:
+                    secondRetrievalStates.purposeEntry.purposeEserviceId,
+                }),
+                agreementId: extractAgreementIdFromAgreementPK(
+                  secondRetrievalStates.agreementEntry.PK
+                ),
+                agreementState: secondRetrievalStates.agreementEntry.state,
+                GSIPK_eserviceId_descriptorId: makeGSIPKEServiceIdDescriptorId({
+                  eserviceId:
+                    secondRetrievalStates.purposeEntry.purposeEserviceId,
+                  descriptorId:
+                    secondRetrievalStates.agreementEntry.agreementDescriptorId,
+                }),
+                descriptorState: secondRetrievalStates.agreementEntry.state,
+                descriptorAudience:
+                  secondRetrievalStates.catalogEntry.descriptorAudience,
+                descriptorVoucherLifespan:
+                  secondRetrievalStates.catalogEntry.descriptorVoucherLifespan,
+                purposeState: secondRetrievalStates.purposeEntry.state,
+                purposeVersionId:
+                  secondRetrievalStates.purposeEntry.purposeVersionId,
+                updatedAt: new Date().toISOString(),
+              };
+
+            await upsertTokenStateClientPurposeEntry(
+              updatedTokenClientPurposeEntry,
+              dynamoDBClient
+            );
+          }
+        }
+      }
+    })
+    .with({ type: "ClientPurposeRemoved" }, async (msg) => {
+      const client = parseClient(msg.data.client, msg.type);
+      const pk = makePlatformStatesClientPK(client.id);
+      const clientEntry = await readClientEntry(pk, dynamoDBClient);
+
+      if (clientEntry) {
+        if (clientEntry.version > msg.version) {
+          return Promise.resolve();
+        } else {
+          const GSIPK_clientId_purposeId = makeGSIPKClientIdPurposeId({
+            clientId: client.id,
+            purposeId: unsafeBrandId(msg.data.purposeId),
+          });
+          if (client.purposes.length > 0) {
+            // platform-states
+            await cleanClientPurposeIdsInPlatformStatesEntry(
+              pk,
+              msg.version,
+              dynamoDBClient
+            );
+
+            // token-generation-states
+            await deleteEntriesWithClientAndPurposeFromTokenGenerationStatesTable(
+              GSIPK_clientId_purposeId,
+              dynamoDBClient
+            );
+          } else {
+            // platform-states
+            await deleteClientEntryFromPlatformStates(pk, dynamoDBClient);
+
+            // token-generation-states
+            await convertEntriesToClientKidInTokenGenerationStates(
+              GSIPK_clientId_purposeId,
+              dynamoDBClient
+            );
```

Comment:
the client should not be removed for this event.
I rephrased the documentation to ease the understanding of this event as
1. if there are no other purposes, for each kid create a `CLIENTKID` record
2. remove all records for the given client and purpose

does it help?

Reply:
Thanks, fixed.